### ISSUE TITLE
TFS SDK Upgrade: TFS material checkout fails if the material destination directory is changed

### ISSUE THREAD

**Issue by rajiesh** [gocd/gocd#2840]
##### Issue Type

- Bug Report

###### Basic environment details

* Go Version: `16.12.0 (4220-d42276500ba2426a9e1cd4c01061e2f773d09edc)`

##### Steps to Reproduce

1. Create a pipeline with TFS material
2. Trigger the pipeline and wait for it to pass
3. Modify the pipeline material - Change the destination directory
4. Trigger the pipeline

##### Expected Results

Pipeline should be pass

##### Actual Results

Pipeline failed with this error
```
Failed while checking out into Working Folder: pipelines/sdf/some-place, Project Path: $/MyFirstProject, Workspace: 4dc964cb4a8be66e9392bd9d6f8f65d115524fb62eb525735709f79c403b57b4, Username: rajiesh.narayanan@gmail.com, Domain: , Root Cause: The path /Users/rajieshn/workspace/functional-tests/target/go-agent-16.12.0/pipelines/sdf/some-place is already mapped in workspace c736e0367edea19ec0d0aabc4aa1dcc82e0c2b52cbc86b8c38f323b8facc932b
```
##### Log snippets

Error log from the agent when the checkout failed
```
2016-11-07 10:42:06,424 [loopThread] ERROR thoughtworks.go.work.DefaultGoPublisher:135 - Failed while checking out into Working Folder: pipelines/sdf/some-place, Project Path: $/MyFirstProject, Workspace: 4dc964cb4a8be66e9392bd9d6f8f65d115524fb62eb525735709f79c403b57b4, Username: rajiesh.narayanan@gmail.com, Domain: , Root Cause: The path /Users/rajieshn/workspace/functional-tests/target/go-agent-16.12.0/pipelines/sdf/some-place is already mapped in workspace c736e0367edea19ec0d0aabc4aa1dcc82e0c2b52cbc86b8c38f323b8facc932b;rajiesh.narayanan@gmail.com.
java.lang.RuntimeException: Failed while checking out into Working Folder: pipelines/sdf/some-place, Project Path: $/MyFirstProject, Workspace: 4dc964cb4a8be66e9392bd9d6f8f65d115524fb62eb525735709f79c403b57b4, Username: rajiesh.narayanan@gmail.com, Domain: , Root Cause: The path /Users/rajieshn/workspace/functional-tests/target/go-agent-16.12.0/pipelines/sdf/some-place is already mapped in workspace c736e0367edea19ec0d0aabc4aa1dcc82e0c2b52cbc86b8c38f323b8facc932b;rajiesh.narayanan@gmail.com.
  at com.thoughtworks.go.domain.materials.tfs.AbstractTfsCommand.checkout(AbstractTfsCommand.java:98)
  at com.thoughtworks.go.tfssdk.TfsSDKCommandTCLAdapter.checkout(TfsSDKCommandTCLAdapter.java:43)
  at com.thoughtworks.go.config.materials.tfs.TfsMaterial.updateTo(TfsMaterial.java:156)
  at com.thoughtworks.go.domain.MaterialRevision.updateTo(MaterialRevision.java:151)
  at com.thoughtworks.go.domain.materials.AbstractMaterialAgent.prepare(AbstractMaterialAgent.java:41)
  at com.thoughtworks.go.remote.work.BuildWork.prepareJob(BuildWork.java:193)
  at com.thoughtworks.go.remote.work.BuildWork.build(BuildWork.java:141)
  at com.thoughtworks.go.remote.work.BuildWork.doWork(BuildWork.java:96)
  at com.thoughtworks.go.agent.JobRunner.run(JobRunner.java:60)
  at com.thoughtworks.go.agent.AgentController.retrieveWork(AgentController.java:229)
  at com.thoughtworks.go.agent.AgentController.rpcLoop(AgentController.java:183)
  at com.thoughtworks.go.agent.AgentController.loop(AgentController.java:173)
  at sun.reflect.GeneratedMethodAccessor10.invoke(Unknown Source)
  at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:606)
  at org.springframework.util.MethodInvoker.invoke(MethodInvoker.java:273)
  at org.springframework.scheduling.support.MethodInvokingRunnable.run(MethodInvokingRunnable.java:65)
  at org.springframework.scheduling.timer.DelegatingTimerTask.run(DelegatingTimerTask.java:70)
  at java.util.TimerThread.mainLoop(Timer.java:555)
  at java.util.TimerThread.run(Timer.java:505)
Caused by: com.microsoft.tfs.core.clients.versioncontrol.exceptions.MappingConflictException: The path /Users/rajieshn/workspace/functional-tests/target/go-agent-16.12.0/pipelines/sdf/some-place is already mapped in workspace c736e0367edea19ec0d0aabc4aa1dcc82e0c2b52cbc86b8c38f323b8facc932b;rajiesh.narayanan@gmail.com.
  at com.microsoft.tfs.core.clients.versioncontrol.workspacecache.internal.InternalCache.checkForMappingConflicts(InternalCache.java:212)
  at com.microsoft.tfs.core.clients.versioncontrol.workspacecache.internal.InternalCache.checkForMappingConflicts(InternalCache.java:176)
  at com.microsoft.tfs.core.clients.versioncontrol.VersionControlClient.updateWorkspace(VersionControlClient.java:1365)
  at com.microsoft.tfs.core.clients.versioncontrol.soapextensions.Workspace.update(Workspace.java:6614)
  at com.microsoft.tfs.core.clients.versioncontrol.soapextensions.Workspace.update(Workspace.java:6567)
  at com.microsoft.tfs.core.clients.versioncontrol.soapextensions.Workspace.update(Workspace.java:6525)
  at com.microsoft.tfs.core.clients.versioncontrol.soapextensions.Workspace.update(Workspace.java:6494)
  at com.microsoft.tfs.core.clients.versioncontrol.soapextensions.Workspace.update(Workspace.java:6470)
  at com.microsoft.tfs.core.clients.versioncontrol.soapextensions.Workspace.update(Workspace.java:6446)
  at com.microsoft.tfs.core.clients.versioncontrol.soapextensions.Workspace.createWorkingFolders(Workspace.java:1971)
  at com.microsoft.tfs.core.clients.versioncontrol.soapextensions.Workspace.createWorkingFolder(Workspace.java:1847)
```

**Comment by alexbestul** [comment]
I've run into this same issue before.  I tried to troubleshoot it by reading the logs and source code, and eventually came to the conclusion that the bug was in the TFS SDK, and not in Go itself.  I never logged a bug, but I did write up my findings to share with my team.  I've included that writeup below, in the hopes that it can help somebody.


------------------------------------


# Symptoms

A pipeline fails with an error like:

```
20:23:27.957 Failed while checking out into Working Folder: pipelines\My_Pipeline_Name\material_directory_name, Project Path: $/TFS/Project/Path/, Workspace: <new_tfs_workspace_name>, Username: tfs_user, Domain: tfs_domain, Root Cause: The working folder C:\go\agent\pipelines\My_Pipeline_Name\material_directory_name is already in use by the workspace <existing_tfs_workspace_name>;tfs_user on computer MY_GO_AGENT.
```

The error is most likely to occur after a change to a TFS material in the pipeline's configuration.  It may also occur after an agent's GUID changes, for example if the agent application was removed and reinstalled.

# Workaround

Use the TF command line tool to delete the offending workspace.  You must have sufficient privileges to run the command--you must either be a TFS server administrator, or the owner of the workspace.

```
tf workspace /delete existing_tfs_workspace_name;tfs_user
```

TF is included with Visual Studio, and should be in one of the following locations, depending on the Visual Studio version installed:

* Visual Studio 2013:  `C:\Program Files (x86)\Microsoft Visual Studio 12.0\Common7\IDE\TF.exe`
* Visual Studio 2015:  `C:\Program Files (x86)\Microsoft Visual Studio 14.0\Common7\IDE\TF.exe`

# Simplified Cause

For any pipeline that uses a TFS material, a Go agent running that pipeline must create a TFS workspace in order to fetch the material.  The workspaces themselves are treated as throwaway, and are un-mapped (but not deleted) immediately after the material is fetched.  In some cases, the workspace is not un-mapped, which may cause subsequent runs of the pipeline to fail.  The exact reason that the un-mapping fails is not known.

In order for the issue to surface, the following things must happen:

* A pipeline must be configured to use a TFS material.
* A prior run of the pipeline must fail to un-map the material's TFS workspace.
* A change must be made to the material's configuration, or to the agent's GUID.

If all of the above conditions are met, all subsequent runs of the pipeline will fail.  It is assumed that the pipeline is running on a single agent.  If multiple agents are used, it may be possible for the issue to only affect specific agents.

# Root Cause

When Go creates a TFS workspace, it uses certain properties of the TFS material, along with the GUID of the Go agent, to create a hash, which is used as the workspace name.  When the material's configuration or the agent's GUID changes, its hash also changes, and a new workspace is created.  If an existing workspace is already mapped to the same location on disk, the TFS SDK throws an exception.

The intended behavior is that the Go agent un-map the workspace immediately after the TFS material is fetched.  That behavior should prevent workspace mapping collisions by ensuring that there are never any existing mappings to collide with.  Based on analysis of Go's behavior, it appears that the TFS SDK will sometimes silently fail to un-map the workspace.  The `unMap()` call ([seen here](https://github.com/gocd/gocd/blob/13e9f67c4d7e07310ecc0cb232ca9d0bd7fa3145/common/src/com/thoughtworks/go/domain/materials/tfs/AbstractTfsCommand.java#L159)) is made and returns without error (as evidenced by the Go agent logs, below), but the workspace is not actually unmapped.

The following log snippet was captured with `log4j.logger.com.thoughtworks.go.domain.materials.tfs=DEBUG` set in the log4j configuration.  Although the logs do not show any errors, the workspace did not un-map successfully.

```
2016-08-25 17:54:04,814 [loopThread] DEBUG domain.materials.tfs.TfsCommandFactory:37 - [TFS SDK] Creating TFS SDK Client
2016-08-25 17:54:04,876 [loopThread] INFO  ws.runtime.client.SOAPService:500 - SOAP method='Connect', status=200, content-length=2883, server-wait=47 ms, parse=0 ms, total=47 ms, throughput=61340 B/s, gzip
2016-08-25 17:54:04,876 [loopThread] DEBUG domain.materials.tfs.AbstractTfsCommand:148 - [TFS] Initializing Workspace e824fa064b029bdc772dabb27899147076f8aa1addcd53939ce91165b442a16f, Working Folder pipelines\test\hooray
2016-08-25 17:54:04,923 [loopThread] INFO  ws.runtime.client.SOAPService:500 - SOAP method='QueryWorkspaces', status=200, content-length=63402, server-wait=31 ms, parse=16 ms, total=47 ms, throughput=1348979 B/s, gzip
2016-08-25 17:54:04,953 [loopThread] INFO  ws.runtime.client.SOAPService:500 - SOAP method='CreateWorkspace', status=200, content-length=1077, server-wait=30 ms, parse=0 ms, total=30 ms, throughput=35900 B/s, gzip
2016-08-25 17:54:05,000 [loopThread] INFO  ws.runtime.client.SOAPService:500 - SOAP method='UpdateWorkspace', status=200, content-length=1183, server-wait=31 ms, parse=0 ms, total=31 ms, throughput=38161 B/s, gzip
2016-08-25 17:54:05,016 [loopThread] DEBUG domain.materials.tfs.AbstractTfsCommand:84 - [TFS] Retrieving Files from Workspace e824fa064b029bdc772dabb27899147076f8aa1addcd53939ce91165b442a16f, Working Folder pipelines\test\hooray, Revision StringRevision[30817] 
2016-08-25 17:54:05,032 [loopThread] INFO  ws.runtime.client.SOAPService:500 - SOAP method='QueryWorkspace', status=200, content-length=1179, server-wait=16 ms, parse=0 ms, total=16 ms, throughput=73688 B/s, gzip
2016-08-25 17:54:05,063 [loopThread] INFO  ws.runtime.client.SOAPService:500 - SOAP method='Get', status=200, content-length=4390, server-wait=31 ms, parse=0 ms, total=31 ms, throughput=141613 B/s, gzip
2016-08-25 17:54:05,204 [loopThread] INFO  ws.runtime.client.SOAPService:500 - SOAP method='UpdateLocalVersion', status=200, content-length=367, server-wait=79 ms, parse=0 ms, total=79 ms, throughput=4646 B/s, gzip
2016-08-25 17:54:05,204 [loopThread] DEBUG domain.materials.tfs.AbstractTfsCommand:156 - [TFS] Unmapping Project Path $/InContext/DevOps/VagrantBoxes/ for Workspace e824fa064b029bdc772dabb27899147076f8aa1addcd53939ce91165b442a16f, User alex.bestul 
2016-08-25 17:54:05,219 [loopThread] INFO  ws.runtime.client.SOAPService:500 - SOAP method='QueryWorkspace', status=200, content-length=1179, server-wait=15 ms, parse=0 ms, total=15 ms, throughput=78600 B/s, gzip
2016-08-25 17:54:05,250 [loopThread] INFO  ws.runtime.client.SOAPService:500 - SOAP method='UpdateWorkspace', status=200, content-length=1183, server-wait=31 ms, parse=0 ms, total=31 ms, throughput=38161 B/s, gzip
```

Since no error is raised, the root cause appears to be a bug in the TFS SDK used by Go.

**Comment by ketan** [comment]
@alexbestul — can you confirm the version of GoCD where you are seeing this?

**Comment by alexbestul** [comment]
I saw the issue with versions 15.3 and 16.3.  I may have also seen it with 15.1 and 15.2--I can't remember for sure.

I haven't kept up with the latest versions, so I can't speak for any more recent versions.

**Comment by reidca** [comment]
After upgrading to GO 16.12 we saw issues with TFS workspaces which prevented any builds from working.

The error was not exactly the same however it is likely related to the problems described here:

"The workspace 1f0137cccf515faa1d9364d3a26b1ce2276103ea4be52a6e66f69ba27f54d33d;mytfsuser does not exist."

This issue has meant we have had to roll back to 16.10 pending investigation.

This issue was not present in the earlier version of the SDK or the code used within GO has changed.
